### ISSUE TITLE
Broken since last April with SoloKeys (v1)

### ISSUE THREAD

**Issue by riastradh** [jcjones/u2f-hid-rs#333]
With a SoloKeys Solo1 (USB-A, v1 firmware), authenticator-rs immediately fails with PinNotSet, leading to a confusing dialog box in Firefox about having to set a PIN.

```
[2024-02-09T22:42:25Z ERROR authenticator::statemachine] Error when determining pinAuth: PinError(PinNotSet)
```

By bisection, I narrowed this down to a commit in the branch that served to fix another problem I reported back in April, #255: https://github.com/mozilla/authenticator-rs/pull/258/commits/ae57ddff34092d5385f1e99ccd386ce2630b0820

Planning to look closer, but I haven't yet looked closely at the code to find out what the problem is -- just recording what I've discovered so far, since apparently this means SoloKeys (at least v1) are completely broken with Firefox and have been since the ctap2-2021 merge (114?).

**Comment by jschanck** [comment]
It would help to have the `authenticatorGetInfo` response. You can run `cargo run --example interactive_management` and at the prompt select `(2) Show full info`. Or alternatively run `RUST_LOG=debug cargo run --example ctap2` and look for a line that starts with "AuthenticatorInfo".

**Comment by riastradh** [comment]
```
Authenticator Info AuthenticatorInfo {
    versions: [
        U2F_V2,
        FIDO_2_0,
    ],
    extensions: [
        "hmac-secret",
    ],
    aaguid: AAGuid(8976631b-d4a0-427f-5773-0ec71c9e0279),
    options: AuthenticatorOptions {
        platform_device: false,
        resident_key: true,
        client_pin: Some(
            false,
        ),
        user_presence: true,
        user_verification: None,
        pin_uv_auth_token: None,
        no_mc_ga_permissions_with_client_pin: None,
        large_blobs: None,
        ep: None,
        bio_enroll: None,
        user_verification_mgmt_preview: None,
        uv_bio_enroll: None,
        authnr_cfg: None,
        uv_acfg: None,
        cred_mgmt: None,
        credential_mgmt_preview: None,
        set_min_pin_length: None,
        make_cred_uv_not_rqd: None,
        always_uv: None,
    },
    max_msg_size: Some(
        1200,
    ),
    pin_protocols: Some(
        [
            1,
        ],
    ),
    max_credential_count_in_list: None,
    max_credential_id_length: None,
    transports: None,
    algorithms: None,
    max_ser_large_blob_array: None,
    force_pin_change: None,
    min_pin_length: None,
    firmware_version: None,
    max_cred_blob_length: None,
    max_rpids_for_set_min_pin_length: None,
    preferred_platform_uv_attempts: None,
    uv_modality: None,
    certifications: None,
    remaining_discoverable_credentials: None,
    vendor_prototype_config_commands: None,
}
```

**Comment by jschanck** [comment]
Does it fail when creating a credential, asserting a credential, or both? How about on https://webauthn.io with both "user verification" and "discoverable credential" set to "discouraged"?

**Comment by riastradh** [comment]
- fails when creating a credential
- did not test asserting a credential
- reproduces on https://webauthn.io with both user verification and discoverable credential set to discouraged

**Comment by riastradh** [comment]
OK, sorry for the alarm -- looks like this has been fixed already, but I was testing in firefox-115esr, not latest firefox, and I realize now that I was initially running my authenticator-rs tests in an older authenticator-rs tree (from some time last summer when I was working on other authenticator-rs bugs) and then I bisected from there instead of from the head of the ctap2-2021 branch today.

In conclusion:

- This is a duplicate of https://github.com/mozilla/authenticator-rs/issues/283.
- The problem was fixed in https://github.com/mozilla/authenticator-rs/pull/285 by commit https://github.com/mozilla/authenticator-rs/pull/285/commits/4b87aca52cf68c0e7772c94103f5ad44200bc0fc which went out in authenticator-rs 0.4.0-alpha.19.
- The problem was fixed in Firefox 118 by updating to authenticator-rs 0.4.0-alpha.19.
- The problem persists in Firefox 115.7.0esr, because that is still on authenticator-rs 0.4.0-alpha.15 (and hasn't had any local patches to fix it as far as I can tell).

It would be nice if this could be fixed in 115esr, but I don't know the policies or procedures around that kind of thing, and in any case it's not a problem in authenticator-rs any more so I'll close this issue.
